compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/InitializersLowering.kt

```diff
@@ -7,84 +7,72 @@ package org.jetbrains.kotlin.backend.common.lower
+import org.jetbrains.kotlin.backend.common.IrElementTransformerVoidWithContext
 import org.jetbrains.kotlin.ir.IrStatement
-import org.jetbrains.kotlin.ir.UNDEFINED_OFFSET
 import org.jetbrains.kotlin.ir.declarations.*
-import org.jetbrains.kotlin.ir.expressions.IrBlock
 import org.jetbrains.kotlin.ir.expressions.IrExpression
 import org.jetbrains.kotlin.ir.expressions.IrInstanceInitializerCall
 import org.jetbrains.kotlin.ir.expressions.IrStatementOriginImpl
 import org.jetbrains.kotlin.ir.expressions.impl.IrBlockImpl
 import org.jetbrains.kotlin.ir.expressions.impl.IrGetValueImpl
 import org.jetbrains.kotlin.ir.expressions.impl.IrSetFieldImpl
 import org.jetbrains.kotlin.ir.util.deepCopyWithSymbols
-import org.jetbrains.kotlin.ir.visitors.IrElementTransformerVoid
 import org.jetbrains.kotlin.ir.visitors.transformChildrenVoid
 
 object SYNTHESIZED_INIT_BLOCK : IrStatementOriginImpl("SYNTHESIZED_INIT_BLOCK")
 
-open class InitializersLowering(
-    context: CommonBackendContext
-) : InitializersLoweringBase(context) {
+open class InitializersLowering(context: CommonBackendContext) : InitializersLoweringBase(context) {
     override fun lower(irClass: IrClass) {
-        val nonStaticDeclarations = getDeclarationsWithNonStaticInitializers(irClass)
-        val instanceInitializerStatements = nonStaticDeclarations.mapNotNull { handleDeclaration(irClass, it) }
-        transformInstanceInitializerCallsInConstructors(irClass, instanceInitializerStatements)
-
-        val anonymousInitializers = nonStaticDeclarations.filterTo(hashSetOf()) { it is IrAnonymousInitializer }
-        irClass.declarations.removeAll(anonymousInitializers)
-    }
-
-    private fun getDeclarationsWithNonStaticInitializers(irClass: IrClass): List<IrDeclaration> =
-        irClass.declarations.filter {
+        val instanceInitializerStatements = extractInitializers(irClass) {
             (it is IrField && !it.isStatic) || (it is IrAnonymousInitializer && !it.isStatic)
         }
+        irClass.transformChildrenVoid(object : IrElementTransformerVoidWithContext() {
+            // Only transform constructors of current class.
+            override fun visitClassNew(declaration: IrClass) = declaration
 
-    private fun transformInstanceInitializerCallsInConstructors(irClass: IrClass, instanceInitializerStatements: List<IrStatement>) {
-        irClass.transformChildrenVoid(object : IrElementTransformerVoid() {
-            override fun visitInstanceInitializerCall(expression: IrInstanceInitializerCall): IrExpression {
-                val copiedBlock =
-                    IrBlockImpl(UNDEFINED_OFFSET, UNDEFINED_OFFSET, context.irBuiltIns.unitType, null, instanceInitializerStatements)
-                        .copy(irClass) as IrBlock
-                return IrBlockImpl(irClass.startOffset, irClass.endOffset, context.irBuiltIns.unitType, null, copiedBlock.statements)
-            }
+            override fun visitSimpleFunction(declaration: IrSimpleFunction) = declaration
```

Comment:
Could you explain why this line is needed?

Reply:
Same reason as `visitClassNew`: to not visit the parts of the tree that don't have any instance initializer calls anyway. The comment above applies to both of them, which would be clearer if there was no empty line, but then IntelliJ would complain about a style guide violation...